code/sonalyze/daemon/perform.go

```diff
@@ -0,0 +1,387 @@
+package daemon
+
+import (
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"log/syslog"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"syscall"
+
+	"go-utils/auth"
+	"go-utils/httpsrv"
+	"go-utils/process"
+)
+
+func (dc *DaemonCommand) RunDaemon(_ io.Reader, _, stderr io.Writer) /* never returns */ {
+	if err := statusStart("jobanalyzer/sonalyzed", stderr); err != nil {
+		fmt.Fprintf(stderr, "FATAL ERROR: Failing to open logger: %s", err.Error())
+		os.Exit(1)
+	}
+
+	// Note "daemon" is not a command here
+	http.HandleFunc("/add", httpAddHandler(dc))
+	http.HandleFunc("/jobs", httpGetHandler(dc, "jobs"))
+	http.HandleFunc("/load", httpGetHandler(dc, "load"))
+	http.HandleFunc("/uptime", httpGetHandler(dc, "uptime"))
+	http.HandleFunc("/profile", httpGetHandler(dc, "profile"))
+	http.HandleFunc("/parse", httpGetHandler(dc, "parse"))
+	http.HandleFunc("/metadata", httpGetHandler(dc, "metadata"))
+	// These request names are compatible with the older `infiltrate` and with the upload infra
+	// already running on the clusters.
+	http.HandleFunc("/sonar-freecsv", httpPostHandler(dc, "sample", "text/csv"))
+	http.HandleFunc("/sysinfo", httpPostHandler(dc, "sysinfo", "application/json"))
+
+	var programFailed bool
+	s := httpsrv.New(dc.Verbose, int(dc.port), func(err error) {
+		programFailed = true
+	})
+	go s.Start()
+
+	// Wait here until we're stopped by SIGHUP (manual) or SIGTERM (from OS during shutdown).
+	//
+	// TODO: IMPROVEME: For SIGHUP, we should not exit but should instead reread the password file
+	// and the cluster aliases file.
+	process.WaitForSignal(syscall.SIGHUP, syscall.SIGTERM)
+	s.Stop()
+
+	if programFailed {
+		os.Exit(1)
+	}
+	os.Exit(0)
```

Comment:
Oops, os.Exit is wrong here (it's a holdover from sonalyzed).  It will prevent the database from being closed properly by the unwind handler in sonalyze.go.  Instead, the this method must return an error, and the caller must handle that error properly.

Reply:
Should search the entire code base for os.Exit(), we don't want these.

Reply:
Ditto things like Fatal / Fatalf.